### PR TITLE
Localization PR

### PR DESCRIPTION
**Pull request by strega-nil**
v exciting

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -121,6 +123,23 @@ endif()
 
 include(GNUInstallDirs)
 
+# ====================
+# === Dependencies ===
+# ====================
+
+include(FetchContent)
+FetchContent_Declare(
```
Comment (ras0219-msft) [inline CMakeLists.txt:131]: I know it's "documenting the standard library", but we should have a comment about how to point this at a pre-cloned copy of fmt.

File: src/vcpkg/base/downloads.cpp [hunk 2]
```diff
@@ -476,13 +482,14 @@ namespace vcpkg::Downloads
     }
 #endif
 
+    DECLARE_AND_REGISTER_MESSAGE(DownloadBasicErrorWithInfo, (msg::url, msg::error), "{LOCKED}", "{url}: {error}");
```
Comment (ras0219-msft) [inline src/vcpkg/base/downloads.cpp:485]: Docs on what `"{LOCKED}"` means?

File: src/vcpkg/base/messages.cpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+                   ? static_cast<DWORD>(-1)
+                   : static_cast<DWORD>(size);
+    }
+
+    void write_unlocalized_text_to_stdout(Color c, StringView sv)
+    {
+        if (sv.empty()) return;
+
+        auto handle = ::GetStdHandle(STD_OUTPUT_HANDLE);
+        if (is_console(handle))
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:39]: This should be cached; we don't want to call `GetConsoleMode()` every time we print a message.

File: src/vcpkg/base/messages.cpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            {
+                DWORD written = 0;
+                check_write(::WriteFile(handle, pointer, size_to_write(size), &written, nullptr));
+                pointer += written;
+                size -= written;
+            }
+        }
+    }
+#else
+    static void write_all(int fd, const char* ptr, size_t to_write)
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:83]:
```suggestion
    static void write_all(const char* ptr, size_t to_write)
```
The body assumes stdout, so it should not take a parameter.

File: src/vcpkg/base/messages.cpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    void write_unlocalized_text_to_stdout(Color c, StringView sv)
+    {
+        static constexpr char reset_color_sequence[] = {'\033', '[', '0', 'm'};
+
+        if (sv.empty()) return;
+
+        bool reset_color = false;
+        if (::isatty(STDOUT_FILENO) && c != Color::None)
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:105]: If we are introducing tty detection, we need to give the user a way to override it (e.g. `--color` or `--no-color`).

File: src/vcpkg/base/messages.cpp [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            write_all(STDOUT_FILENO, reset_color_sequence, sizeof(reset_color_sequence));
+        }
+    }
+#endif
+
+    namespace
+    {
+        struct Messages
+        {
+            // this is basically a SoA - each index is:
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:126]:
```suggestion
            // this is an SoA - each index is:
```

File: src/vcpkg/base/messages.cpp [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+            // {
+            //   name
+            //   default_string
+            //   localization_comment
+            //   localized_string
+            // }
+            // requires: names.size() == default_strings.size() == localized_strings.size()
+            std::vector<StringLiteral> names;
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:134]:
```suggestion
            std::vector<StringLiteral> names;  // const after startup
```

File: src/vcpkg/base/messages.cpp [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    }
+
+    void threadunsafe_initialize_context()
+    {
+        Messages& m = messages();
+        if (m.initialized)
+        {
+            write_unlocalized_text_to_stdout(Color::Error,
+                                 "double-initialized message context; this is a very serious bug in vcpkg\n");
+            Checks::exit_fail(VCPKG_LINE_INFO);
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:158]:
```suggestion
            Checks::exit_with_message(VCPKG_LINE_INFO, "Attempted to double-initialized message context; this is a vcpkg internal bug");
```

File: src/vcpkg/base/messages.cpp [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+
+        Messages& m = messages();
+        std::vector<std::string> names_without_localization;
+
+        for (::size_t index = 0; index < m.names.size(); ++index)
+        {
+            const auto& name = m.names[index];
+            if (auto p = message_map.get(m.names[index]))
+            {
+                m.localized_strings[index] = p->string().to_string();
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:175]: Needs error handling of `p` not being a string.

File: src/vcpkg/base/messages.cpp [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+            }
+            else if (Debug::g_debugging)
+            {
+                // we only want to print these in debug
+                names_without_localization.push_back(name);
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:180]: I don't think `--debug` should immediately imply "never localize anything". Maybe there is a need for a separate `--debug-locale` or some other method of getting this output.

File: include/vcpkg/base/messages.h [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+        // REQUIRES: index < last_message_index()
+        StringView get_message_name(::size_t index);
+        // REQUIRES: index < last_message_index()
+        StringView get_default_format_string(::size_t index);
+        // REQUIRES: index < last_message_index()
+        StringView get_localization_comment(::size_t index);
+    }
+
+    // load from "locale_base/${language}.json"
+    void threadunsafe_initialize_context(const Filesystem& fs, StringView language, const Path& locale_base);
```
Comment (ras0219-msft) [inline include/vcpkg/base/messages.h:75]: Noting that the user could set language to `C:\Users\My User\foo` and get that loaded as a language file.

File: src/vcpkg/base/messages.cpp [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+        threadunsafe_initialize_context(message_map.first.object());
+    }
+
+    ::size_t detail::number_of_messages() { return messages().names.size(); }
+
+    ::size_t detail::startup_register_message(StringLiteral name, StringLiteral format_string, StringLiteral comment)
+    {
+        Messages& m = messages();
+
+        auto name_is_used = std::find(m.names.begin(), m.names.end(), name);
```
Comment (ras0219-msft) [inline src/vcpkg/base/messages.cpp:217]: This is `O(N^2)` -- we should not be doing `O(N^2)` things on every startup. Instead, let's only check this when generating the localization file (which will be part of CI).

File: src/vcpkg/base/system.cpp [hunk 13]
```diff
@@ -372,4 +372,22 @@ namespace vcpkg
...
 {
     std::atomic<bool> g_debugging(false);
+
+    void println(StringView sv)
+    {
+        if (g_debugging)
+        {
+            msg::write_text_to_stdout(Color::None, "[DEBUG] ");
+            msg::write_text_to_stdout(Color::None, sv);
+            msg::write_text_to_stdout(Color::None, "\n");
```
Comment (ras0219-msft) [inline src/vcpkg/base/system.cpp:382]:
```suggestion
            msg::write_text_to_stdout(Color::None, Strings::concat("[DEBUG] ", sv, "\n");
```

Message printing performance on Windows is _atrocious_ and in `--debug` mode combining these print calls will run significantly faster.

File: src/vcpkg/base/system.cpp [hunk 14]
```diff
@@ -372,4 +372,22 @@ namespace vcpkg
...
+        }
+    }
+    void print(StringView sv)
+    {
+        if (g_debugging)
+        {
+            msg::write_text_to_stdout(Color::None, "[DEBUG] ");
```
Comment (ras0219-msft) [inline src/vcpkg/base/system.cpp:389]: Ditto as above.

File: src/vcpkg/binarycaching.cpp [hunk 15]
```diff
@@ -1360,6 +1367,46 @@ namespace vcpkg
 
 namespace
 {
+    DECLARE_AND_REGISTER_MESSAGE(BcPathsMustBeAbsolute, (), "", "expected arguments: path arguments for binary config strings must be absolute");
```
Comment (ras0219-msft) [inline src/vcpkg/binarycaching.cpp:1370]: This really makes me wonder whether the regex based solutions are better; this does seems pretty awful for code review and maintenance being so far out of context.

File: src/vcpkg/base/messages.cpp (hunk 10 above)
Comment (strega-nil) [reply src/vcpkg/base/messages.cpp:180]: this is a misunderstanding of how this works.

File: src/vcpkg/binarycaching.cpp (hunk 15 above)
Comment (strega-nil) [reply src/vcpkg/binarycaching.cpp:1370]: I am not going to implement regex based solutions. I think it would be a major mistake.

**Pull request closed without merging**